qcodes/instrument/abstract.py

```diff
@@ -0,0 +1,101 @@
+from typing import Any, Type
+from qcodes import Parameter
+
+
+class AbstractParameter(Parameter):
+
+    """
+    This is a trivial subclass of 'Parameter' to signal
+    that this parameters *must* be overridden in
+    subclasses of abstract instruments.
+    """
+
+
+class AbstractParameterException(Exception):
+    """
+    Errors with abstract parameter utilization should
+    raise this exception class.
+    """
+
+
+def abstract_instrument(cls: Type) -> Type:
+    """
+    A class decorator to create an abstract instrument. Abstract
+    instruments are allowed to have abstract parameters, but
+    their subclasses do not. This works by replacing the
+    'add_parameter' and '__init_subclass__' methods of the class
+    dynamically.
+
+    Args:
+        cls: The class to be decorated
+
+    Returns:
+        The decorated class
+    """
+
+    def __init_subclass__(sub_cls: Type) -> None:
+
+        __init__ = sub_cls.__init__
+
+        def __init_new__(self, *args: Any, **kwargs: Any) -> None:
+            """
+            Subclasses of an abstract instrument should check
+            after initialization whether there still are
+            abstract parameters. If there are, we should raise
+            an exception.
+            """
+            __init__(self, *args, **kwargs)
+            # after the usual initialization...
+
+            abstract_parameters = [
+                parameter.name for parameter in self.parameters.values()
+                if isinstance(parameter, AbstractParameter)
+            ]
+
+            if any(abstract_parameters):
+                cls_name = sub_cls.__name__
+
+                raise AbstractParameterException(
+                    f"Class '{cls_name}' has un-implemented Abstract Parameter(s): " +
+                    ", ".join([f"'{name}'" for name in abstract_parameters])
+                )
+
+        sub_cls.__init__ = __init_new__
+
+    original_add_parameter = cls.add_parameter
+
+    def add_parameter(
+            self, name: str, parameter_class: type = Parameter,
```

Comment:
to indicate that parameter_class is expected to be a subclass of _BaseParameter (as in the original method (i hope) )
```suggestion
            self, name: str, parameter_class: Type[_BaseParameter] = Parameter,
```

Reply:
Is the leading underscore a typo?